### ISSUE TITLE
Printing debug messages from in-game?

### ISSUE THREAD

**Issue by Fireboyd78** [hrydgard/ppsspp#10716]
I'm trying to have a look at debug information that can be printed out from Driver '76. They have a debug logging function that uses the same format as `printf`, however it's been "disabled" by not actually doing anything with the arguments and simply adds 1 to a global counter of some sort.

I can find these functions in IDA & PPSSPP quite easily, and attempted to redirect flow directly to the `printf` function in the executable itself. However, this results in an error being spammed many many times:

```
user_main E[SCEIO]: HLE\sceIo.cpp:1065 sceIoWrite ERROR: no file open
```

So to clarify, I'm taking this function:
```
RIDebugLog:
	addiu	sp,sp,-0x20
	lui	a0,0x8C0
	lw	a1,0x19FC(a0)
	addiu	a1,a1,0x1
	sw	a1,0x19FC(a0)
	jr	ra
	addiu	sp,sp,0x20
```
And using PPSSPP's handy disassembler, changing it to this:
```
RIDebugLog:
	j printf
	nop
	# ... etc
```
So in essence, I'm redirecting `RIDebugLog` calls to `printf` (the one in the game executable, not PPSSPP). But of course it wasn't that easy and didn't work as expected.

**TLDR**;
 - Want to redirect debug logging in Driver '76 to PPSSPP's console
 - Game's debug logger is "disabled", but has same prototype as a call to `printf`
 - Tried jumping directly to `printf` in game binary, but results in console being spammed with error messages
 - Was hoping that `printf` would print to stdout and thus PPSSPP's console

Is this possible to do in PPSSPP without any modifications? I didn't want to dedicate too much time to this, but it would be pretty cool to achieve nonetheless. It'd be very helpful in getting filenames the game loads, since they're packed into an archive and referenced by their hashes.

Any help is appreciated!

**Comment by Fireboyd78** [comment]
It appears `sceKernelPrintf` is very close to what I need, unfortunately it's not implemented very well and throws a lot of `UNIMPL` errors. So while technically a "solution", I don't think it's the correct one. I still don't know why `printf` isn't working (stdout not initialized?)

**Comment by hrydgard** [comment]
It's quite clear that when the developers disabled the debug printing for the release executable of the game, they also don't bother opening stdout (or whatever file printf was printing to before) anymore, so that can't work. You can of course hack PPSSPP's sceIoWrite implementation to output data to the console directly if the passed file handle is invalid. Alternatively, you could just add your own HLE logprint function to PPSSPP and insert a sc instruction to call it. I don't think there's a way to do this without hacking PPSSPP at this point...

**Comment by Fireboyd78** [comment]
@hrydgard Thank you so much, I suppose the easiest way to do this would be to hack `sceIoWrite` like you said. I'd be interested in knowing more about creating a custom HLE logprint function, but looking at `sceKernelPrintf`'s implementation it's expecting a string, yet technically should have variadic args. Which leads me to believe those type of functions aren't supported via the HLE system, so I'd need to know how that would work.

I'll see what I can do about hacking `sceIoWrite` like you said. Thanks!

**Comment by unknownbrackets** [comment]
Actually, it's not necessary to open a file if you use `1` (sdtdout) or `2` (stderr) as the fd.

```c
sceIoWrite(2, "foo\n", 4);
```

Should write "foo" to the log.  This is handled here:
https://github.com/hrydgard/ppsspp/blob/dc6a4ea753dd7266807c0ba5dbbba5964492aa39/Core/HLE/sceIo.cpp#L1015-L1021

I'm not sure where printf() is writing to if not stdout/stderr?  It must be passing a file handle either == 0 or > 2, right?

-[Unknown]

**Comment by Fireboyd78** [comment]
So after hours of debugging, it appears that `sceKernelPrintf` is incorrect in that it claims to have an `int` return type (it's supposed to be `void`)...I guess the use of this function is incredibly rare, because it will definitely cause crashes when `v0` is not preserved between calls!

**Comment by hrydgard** [comment]
In the MIPS ABI, I don't think there is any requirement that void functions preserve the contents of v0/v1. So I guess that whatever function used to be called before you patched it to sceKernelPrintf simply did preserve it by accident?

**Comment by Fireboyd78** [comment]
@hrydgard I'm guessing it's more of a compiler optimization -- the compiler assumes that `v0` will not be dirty after making a call to a function that returns `void`. If it does, it will stash it in a temporary variable before calling it and restore it afterwards. Since we're obviously not recompiling the source code for games here, it's best not to have functions that should be `void` messing up the `v0` register.

At the moment I can't actually say for sure if changing `sceKernelPrintf` to a void return function worked or not. But preserving `v0` after the call prior to the code fix worked fine, so I'm assuming it fixed the need for that.

**Comment by unknownbrackets** [comment]
Pretty much every PSP firmware function (not including library functions like PGF, etc.) will reset all temporary registers to 0xDEADBEEF.  That setting isn't some optimization or something - it's there to behave like the official firmware does.  This is presumably for kernel security.

If you're looking at the community pspsdk.h to make this assertion, please be aware it's wrong in many places.

ABIs on all platforms - including x86 - don't work the way you've described.  After a function call, it's certainly not safe to expect EAX to be preserved, whether that function call is void or not.

If there is a problem with v0, it may be that sceKernelPrintf is supposed to (sometimes?) return a different value, which is much more likely than it preserving at, v0-v1, a0-a3, or t0-t7 from before the call.  Those registers are, by definition, temporaries.

If you want to create a test (see the pspautotests repo) to validate the behavior, I'd be happy to be proven wrong.

-[Unknown]

**Comment by Fireboyd78** [comment]
Thanks for the heads up, I think the DEADBEEF "fix" was me trying to prevent the game from crashing when logging stuff.

My _assumption_ is that sceKernelPrintf should be a void return type that doesn't set DEADBEEF registers, because there may have been developers that made use of it. Or I could be completely wrong and it turns out to be an internal function not exposed to normal users. I trust the better judgement of you guys since you maintain a working PSP emulator than I ever could. If this really is a problem I'll double check if there's anything I can help fix, but otherwise I guess this issue is closed.

**Comment by unknownbrackets** [comment]
Just remember that in assembly, there's basically only a set number of variables.  MIPS has 31.

One option would be, a void function would need to preserve any of its callers variables.  This makes sense, but it means that the void function has to do a lot of work to save all those variables when it wants to use them.  This slows down programs.

Another option would be, no matter what function you call, all your variables get trashed.  But then you get the problem the other way: the function caller always has to save all its variables, and this is slow.

Both are bad, so we use a compromise.  This is why "ABIs" or "application binary interfaces" exist.  They're used on all architectures: MIPS (like the PSP/PS1/etc.), ARM (probably your phone), x86 (most desktops), etc.

The compromise is: you say some registers are "callee-saved" and some are "caller-saved".  If you play your cards right, no one has to save variables (registers) at all.

Consider this example MIPS function:

```asm
add t0, a0, a0 ; Meaning: t0 = a0 * 2
sub t0, a1 ; Meaning: t0 -= a1
sw a0, 0(a2) ; Meaning: *a2 = t0;
jr ra
```

Which is equivalent to C code:

```c
void foo(int a, int b, int *c) {
    *c = (a * 2) - b;
}
```

You'll notice that it used `t0` as a temporary variable.  This is because t0 is caller-saved.  It's totally safe to destroy all of the t0-t7 registers, and use them for anything.  Anyone using the ABI promises that they'll never leave anything important in caller-saved registers before a function call.

Most of the MIPS registers are caller-saved.  Only 8 (s0-s7) are callee-saved, which means the function must preserve them.  S stands for "saved".

Importantly, v0 and v1 (V is for "value") are not callee-saved.  For a function returning int, v0 is the return value and v1 is just a temporary.  For a void function, both are temporaries.  For a function returning int64_t, both are the return value.  That's just how the ABI defines them.

It is possible for a PSP game to, internally, use its own ABI.  Just like in PPSSPP: our jit uses a custom ABI to run faster.  However, all of the syscalls, kernel or otherwise, use the standard MIPS ABI I've described above.

With that said - sometimes even a void function deterministically (but "accidentally") leaves v0 at a specific value.  Or maybe (this is pretty darn unlikely) it happens to accidentally preserve it.  No C compiler depends on this, with one exception:

Uninitialized values.

So imagine this code:

```c
int bar(int a) {
   int b;
   return a + b;
}
```

Obviously, the above code is incorrect - `b` could be anything.  However, it's possible that it accidentally always has a consistent value, depending on what register the compiler allocated for `b` and what functions and code runs before `bar()` is called.

So what you're probably seeing is that you're calling a function without using the ABI correctly.

-[Unknown]

**Comment by junelva** [comment]
In my case I was working on some homebrew and this error occured when I had symlinked my PSP data directory instead of copying it into the flatpak release's folder. Maintaining and working with multiple installs of PPSSPP is  somewhat inconvenient without the ability to change where that folder is placed

```bash
user_main E[SCEIO]: HLE\sceIo.cpp:1065 sceIoWrite ERROR: no file open
```

This was fixed by not symlinking the PSP user directory from a prior install
